### ISSUE TITLE
Initialize OpenFGA with a predefined StoreID

### ISSUE THREAD

**Issue by ashok-an** [openfga/openfga#1132]
#### Why?
Like seeding DB engines with predefined DBs, there is a need to start a openFGA with predefined storeIDs (any valid ULID)

#### What is expected?
Run OpenFGA container with a seeding StoreID, a store should get created with the given ULID and appropriate log message displayed for verification

**Comment by jon-whit** [comment]
@ashok-an could you provide some examples of shell commands demonstrating what the OpenFGA startup process would look like with this?

What would an ideal cli interface look like to achieve this?

**Comment by ashok-an** [comment]
> @ashok-an could you provide some examples of shell commands demonstrating what the OpenFGA startup process would look like with this?
> 
> What would an ideal cli interface look like to achieve this?

For simplicity, how it looks today is:
```
❯ fga store create --name "store-A"
{
  "store": {
    "created_at":"2023-11-08T04:12:09.966412585Z",
    "id":"01HEPJ2TBDBZEAMEWSGMP95SZ9",
    "name":"store-A",
    "updated_at":"2023-11-08T04:12:09.966412585Z"
  }
}

❯ fga store create --name "store-A"
{
  "store": {
    "created_at":"2023-11-08T04:12:14.882152171Z",
    "id":"01HEPJ2Z52WP8RK2G5CDNJ4MW5",
    "name":"store-A",
    "updated_at":"2023-11-08T04:12:14.882152171Z"
  }
}
```

What I mean/look for is:
❯ fga store create --name "store-A"
{
  "store": {
    "created_at":"2023-11-08T04:12:09.966412585Z",
    "id":"01HEPJ2TBDBZEAMEWSGMP95SZ9",
    "name":"store-A",
    "updated_at":"2023-11-08T04:12:09.966412585Z"
  }
}

❯ fga store create --name "store-A"
{
  "store": {
    "created_at":"2023-11-08T04:12:14.882152171Z", // original creation timestamp
    "id":"01HEPJ2Z52WP8RK2G5CDNJ4MW5", // original ID - 01HEPJ2TBDBZEAMEWSGMP95SZ9
    "name":"store-A",
    "updated_at":"2023-11-08T04:12:14.882152171Z"
  }
}

**Comment by jon-whit** [comment]
@ashok-an I think you may have accidentally copy/pasted the same thing twice.  The two examples (before/after) are identical.

What I may infer from what you're trying to do, just based on the description you have, is that it sounds like you want an idempotent `CreateStore` method that operates on a Store name (e.g. only a single store of a given name can exist in OpenFGA). Is that what you mean?

**Comment by ashok-an** [comment]
Yes!
-- Ashok

**Comment by fsedano** [comment]
@jon-whit @ashok-an I updated an existing issue for this same use case:

https://github.com/openfga/openfga/issues/989

Thanks,

**Comment by PatrickMunsey** [comment]
I just posted an issue in the cli repo https://github.com/openfga/cli/issues/331#issue-2311781366 prior to coming accross this issue here. 

Allowing store-id to be specified would be really benifitial in simplifying deployment configuration. The goal would be to have the entire openfga deployment and configuraiton being able to be achieved repeatably with pre defined values. Requiring randomly generated `store-id` that changes during each deployment to be used for cli connections is cumbersome as opposed to something like a db connection string.

Some examples of how the usage could look:
`fga store import --file test-store.yaml --store-id 01HYHGP755KA1BQMBR568RVKKV`
or
`fga store create --name test-store --model test-model.fga --store-id 01HYHGP755KA1BQMBR568RVKKV`

I foresee being able to specify model-id in a similar fation would go a long way as well.

**Comment by danielloader** [comment]
Yeah this would be quite useful, the alternative I've taken to is just using pg_dump and moving around a database instead and frankly it's not ideal.

For local dev/CI purposes being able to define the model id and the store id feel almost like a requirement for a docker compose stack.

**Comment by abdok96** [comment]
This is very essential when using docker compose and setting up some init scripts. 
There should be a way to determine the store id so that it can be passed as env variables to other services.

**Comment by lvieirajr** [comment]
+1 to being able to set static model and store ids

**Comment by c-thiel** [comment]
+1 for this.

If we cannot set the store-id on store creation, then the store-id becomes dynamic state that needs to be stored.
This is super inconvenient for stateless docker or K8s setups.

Imagine a "migration" container that runs before the app and is supposed to set an auth model (schema) to openfga.

The first time it runs, it currently creates the store and gets a random ULID in return. It then has to pass this on to the application, for example via an emptyDir, which is already not nice.

The second time this container runs, it must determine which migration to apply next. As the container is stateless, it does not even know which stores exist. It has to paginate through all stores until the store name matches. If we could specify the store-id on creation, this wouldn't be required as the store-id becomes easy to manage static information.

**Comment by flex-seongbok** [comment]
How about making store name unique and search store by name?
I think what we need is reproducable and expectable environment.

**Comment by danielloader** [comment]
_**Disclaimer**: I hate issue bumping because of the notification spam - but making an exception here._

Any chance this could be implemented? Is it a case of needing a PR or a case of fundamentally not wanting to have some static configuration options? 

My team has gone as far as just mocking OpenFGA in our golang test suites now instead of using OpenFGA containers in pipeline tests - which to me feels like adding additional risks maintaining a mock when OpenFGA in memory/sqlite mode is effectively ideal for CI testing.

**Comment by miparnisari** [comment]
Hi folks, catching up with this conversation so forgive any misunderstandings on my part.

> How about making store name unique and search store by name?

The second part is done - `ListStores` accepts a `name` parameter. However, the first part isn't guaranteed by neither MySQL or Postgres at the moment, and making it so would be a backwards incompatible change (but maybe this is okay for most users).

If we implemented that guarantee, we could introduce a configuration option in OpenFGA so that calls to `CreateStore(name)` return 409 if a store with that name already exists.

**Comment by seandlg** [comment]
Until this is addressed: What I tend to do, is set up dev OpenFGA as follows using `docker compose`:

```yaml
services:

  openfga-dev:
    image: openfga/openfga:latest
    container_name: openfga-dev
    command: run
    environment:
      - OPENFGA_AUTHN_METHOD=preshared
      - OPENFGA_AUTHN_PRESHARED_KEYS=secret_key_changeme
    ports:
      - "8080:8080"

  openfga-init:
    container_name: openfga-init
    build:
      context: .
      dockerfile: openfga/fga.Dockerfile
    volumes:
      - ./openfga/model.fga.yaml:/model.fga.yaml
      - store_id:/store_id
      - model_id:/model_id
    depends_on:
      openfga-dev:
        condition: service_healthy

volumes:
  store_id:
  model_id:
```

with `./openfga/fga.Dockerfile`:

```Dockerfile
FROM golang:alpine AS builder

# Install FGA CLI using go install
RUN go install github.com/openfga/cli/cmd/fga@latest

FROM alpine:latest

# Install required packages
RUN apk add --no-cache bash curl jq

# Copy FGA binary from builder stage
COPY --from=builder /go/bin/fga /usr/local/bin/fga
COPY ./openfga/setup.sh /setup.sh
RUN chmod +x /setup.sh

ENTRYPOINT ["/setup.sh"]
```

and `./openfga/setup.sh`

```bash
#!/bin/ash

echo "Checking for existing stores..."
export FGA_API_URL=http://openfga-dev:8080
export FGA_API_TOKEN=secret_key_changeme

# Get or create store
FGA_STORE_ID=$(fga store list | jq -r ".stores | sort_by(.created_at) | last | .id")
if [ -z "$FGA_STORE_ID" ] || [ "$FGA_STORE_ID" = "null" ]; then
  echo "Creating new store..."
  fga store import --file=/model.fga.yaml
  FGA_STORE_ID=$(fga store list | jq -r ".stores | sort_by(.created_at) | last | .id")
fi

echo "$FGA_STORE_ID" > /store_id/id
export FGA_STORE_ID=$(cat /store_id/id)
echo "Using store ID: $FGA_STORE_ID"

# Get model ID
FGA_MODEL_ID=$(fga model list --store-id "$FGA_STORE_ID" | jq -r ".authorization_models | sort_by(.created_at) | last | .id")
if [ -z "$FGA_MODEL_ID" ] || [ "$FGA_MODEL_ID" = "null" ]; then
  echo "Error: No OpenFGA model found"
  exit 1
fi

echo "$FGA_MODEL_ID" > /model_id/id
echo "Using model ID: $FGA_MODEL_ID"
```

Make sure to add your `./openfga/model.fga.yaml`, see the [examples](https://github.com/openfga/cli/tree/main/example) for reference. Services can now consume the latest store/model ID via the Docker volume.

Then prep the `fga` CLI using:

```bash
export FGA_API_URL='http://localhost:8080' && export FGA_API_TOKEN='secret_key_changeme' && export FGA_STORE_ID=$(fga store list | jq -r '.stores | sort_by(.created_at) | last | .id') && export FGA_MODEL_ID=$(fga model list | jq -r '.authorization_models | sort_by(.created_at) | last | .id') # PREPFGA
```

In summary, I agree that some guidelines / tooling for a dev setup would be very helpful.
